exonum/src/runtime/mod.rs

```diff
@@ -227,85 +228,93 @@ impl From<RuntimeIdentifier> for u32 {
+    /// This method is called *once* for a specific artifact during the `Runtime` lifetime:
     ///
-    /// - For newly added artifacts, the method is called as the decision to deploy the artifact
-    ///   is made by the supervisor service.
-    /// - After a node restart, the method is called for all previously deployed artifacts.
+    /// - For newly added artifacts, the method is called as the supervisor service decides to deploy
+    ///   the artifact.
+    /// - After the node restart, the method is called for all the previously deployed artifacts.
     // TODO: Elaborate constraints on `Runtime::deploy_artifact` futures (ECR-3840)
     fn deploy_artifact(
         &mut self,
         artifact: ArtifactId,
         deploy_spec: Vec<u8>,
     ) -> Box<dyn Future<Item = (), Error = ExecutionError>>;
 
-    /// Return true if the specified artifact is deployed in this runtime.
+    /// Returns `true` if the specified artifact is deployed in this runtime.
     fn is_artifact_deployed(&self, id: &ArtifactId) -> bool;
 
     /// Runs the constructor of a new service instance with the given specification
     /// and initial arguments. The constructor can initialize the storage of the service,
     /// check for dependencies, etc.
     ///
-    /// The constructor is run *exactly once* during blockchain lifetime for each successfully
-    /// initialized service instance. That is to say, it is *not* called on a node restart.
-    /// The caveat here is "successfully initialized"; at the point `start_adding_service` is called,
-    /// the service is not guaranteed to eventually be added to the blockchain via `commit_service`.
-    /// Indeed, committing the service will not follow if the alternative block proposal without
-    /// the service instantiation was accepted. If the `commit_service` call *is* performed, it is
-    /// guaranteed to be performed in the closest committed block, i.e., before the nearest
-    /// `Runtime::after_commit()`.
-    ///
-    /// The dispatcher does not route transactions and `before_transactions` / `after_transactions`
-    /// events to the service until after `commit_service()` is called with the same instance spec.
-    ///
-    /// The runtime should discard the instantiated service instance after completing this method,
-    /// unless there are compelling reasons to retain it (e.g., creating an instance takes very
-    /// long time).
-    /// Alternatively, "garbage" services may be removed from `Runtime` in `after_commit`
+    /// The constructor runs *exactly once* during the blockchain lifetime for each successfully
+    /// initialized service instance. That is to say, the constructor is *not* called on a node restart.
+    ///
+    /// At the same time, when `start_adding_service` is called,
+    /// there is no guarantee that the service will eventually get to the blockchain via `commit_service`.
+    /// If the consensus accepts an alternative block proposal without the service inside,
```

Comment:
Not sure that service being "inside a proposal" is the best wording. IMO, something like this would be better:

> The consensus may accept an alternative block proposal, in which the service is not instantiated or instantiated with different parameters.